Add ExecutingResponseCreator to delegate request and response

simonbasle opened the pull request:
This commit adds a new convenience `ResponseCreator` implementation that
delegates to a `ClientHttpRequestFactory` to perform a request and
use the response.

For further convenience, a factory method based on a `RestTemplate` is
also added: `MockRestResponseCreators#byExecutingRequestUsing`. This
acts as a pseudo-DSL for fluently setting the ResponseCreator in a
`ResponseActions`.

The primary purpose of this implementation is to perform an actual
request to an external service and use the response in MockMVC.

Closes gh-26381

simonbasle committed:
Add MockRestServiceServer.andPerformRequest

This commit adds the andPerformRequest method to ResponseActions, which
is in effect activating a ResponseCreator that has been implicitly
defined for relevant implementations.

Implementations without explicit support for this will throw an
UnsupportedOperationException by default.

The SimpleRequestExpectationManager and UnorderedRequestExpectationManager
created in `MockRestServiceServerBuilder#build()` capture and use the
original ClientHttpRequestFactory, setting up an implicit ResponseCreator
that executes an actual request and returns the response.

Fixes gh-26381.

rstoyanchev reviewed (changes requested):
Thanks for putting this together based on the discussion under #26381. Looking at the result, I'm wondering if it wouldn't be more idiomatic to simply provide a `ResponseCreator` implementation instead:
```java
public class ExecutingResponseCreator implements ResponseCreator {

	private final ClientHttpRequestFactory requestFactory;


	public ExecutingResponseCreator(RestTemplate restTemplate) {
		this.requestFactory = restTemplate.getRequestFactory();
	}


	@Override
	public ClientHttpResponse createResponse(ClientHttpRequest request) throws IOException {
		Assert.state(request instanceof MockClientHttpRequest, "Expected mock request");
		MockClientHttpRequest mockRequest = (MockClientHttpRequest) request;
		ClientHttpRequest newRequest = this.requestFactory.createRequest(mockRequest.getURI(), mockRequest.getMethod());
		newRequest.getHeaders().putAll(mockRequest.getHeaders());
		StreamUtils.copy(mockRequest.getBodyAsBytes(), newRequest.getBody());
		return newRequest.execute();
	}
}
```

which is then used:
```java
ResponseCreator executingResponseCreator = new ExecutingResponseCreator(restTemplate);
MockRestServiceServer server = MockRestServiceServer.bindTo(restTemplate).build();
...
server.expect(requestTo("/foo")).andRespond(withSuccess());
server.expect(requestTo("/bar")).andRespond(executingResponseCreator);
```

The rest of the code, `MockRestServiceServer`, `ResponseActions`, `RequestExpectationManager` implementations would remain unaware of any of this, which simplifies the implementation, and is generally easier to understand how it fits.

It would provide extra flexibility too because it becomes just another `ResponseCreator` that's no different from the rest. I'm thinking, we could provide an option for setting a default `ResponseCreator` in the `MockRestServiceServer` builder. That's independently useful, but if set to the `ExecutingResponseCreator` it would execute requests by default when `andResponse` isn't called. That's the effect that was sought in the original comment under #26381.

simonbasle committed:
Revert "Add MockRestServiceServer.andPerformRequest"

This reverts commit 3a334a52a963cbb23640c6d5a18227a68a3eb55e.

simonbasle committed:
Add ExecutingResponseCreator to delegate request and response

This commit adds a new convenience `ResponseCreator` implementation that
delegates to a `ClientHttpRequestFactory` to perform a request and
use the response.

For further convenience, a factory method based on a `RestTemplate` is
also added: `MockRestResponseCreators#byExecutingRequestUsing`. This
acts as a pseudo-DSL for fluently setting the ResponseCreator in a
`ResponseActions`.

The primary purpose of this implementation is to perform an actual
request to an external service and use the response in MockMVC.

Closes gh-26381

simonbasle commented:
@rstoyanchev I have reverted the previous implementation and introduced `ExecutingResponseCreator` instead per your suggestion.

simonbasle commented:
I was wondering how to make that feature more discoverable, hence the additional factory method in `MockRestResponseCreators`. Perhaps some mention in documentation is needed as well?

simonbasle commented:
mmh I realize just now that the factory method in `MockRestResponseCreators` doesn't make much sense, because the `ExecutingResponseCreator` MUST be instantiated early on, before the `RestTemplate`'s factory is changed...

I wonder how to improve the discoverability of the feature 🤔

simonbasle committed:
remove factory method, provide usage example in javadoc

simonbasle committed:
fix checkstyle errors

rstoyanchev commented on spring-test/src/main/java/org/springframework/test/web/client/response/ExecutingResponseCreator.java:
```diff
@@ -0,0 +1,71 @@
+
+/**
+ * A {@code ResponseCreator} which delegates to a {@link ClientHttpRequestFactory}
+ * to perform the request and return the associated response.
```
I think it would be good to have a second sentence that starts with "This is useful when...". As far as I can tell it's about testing code that calls multiple remote service, where some need should be called rather than mocked.

simonbasle replied:
thanks for the feedback. I've addressed the documentation and discoverability suggestions.

rstoyanchev reviewed (changes requested):
Thanks that looks very nice and simple.

For the visibility, we could:
-  add a paragraph in the Javadoc for `MockRestResponseCreator` that links to the alternative `ExecutingResponseCreator` along with an `@see` reference.
- add something in the reference docs under [this section](https://docs.spring.io/spring-framework/docs/current/reference/html/testing.html#spring-mvc-test-client).
- yet another option is to add a sample test to the ones [mentioned here](https://docs.spring.io/spring-framework/docs/current/reference/html/testing.html#spring-mvc-test-client-resources), and that could use `MockWebServer` like in `WebClientIntegrationTests` for example

simonbasle committed:
improve documentation and discoverability of feature
